### PR TITLE
Fixes #13668 - adds update task.

### PR DESCRIPTION
**Pull request by ShimShtein**

### CODE REVIEW

File: app/lib/foreman_deployments/tasks/search_task_definition.rb [hunk 1]
```diff
@@ -7,6 +7,26 @@ def run
           results = SerializableArray.new(results)
           SearchTaskDefinition.create_output(results, output)
         end
+
+        def self.dereference_output(storage, path)
```
Comment (tstrachota) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: This method is quite difficult to understand without any context. Would you mind adding some tests to make the purpose more obvious?

File: lib/foreman_deployments/dynflow_output_patches.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+          super
+        end
+      end
+    end
+  end
+
+  class ExecutionPlan::OutputReference
+    prepend Dynflow::ForemanDeploymentsExtensions::OutputReferenceExtensions
+  end
+end
```
Comment (tstrachota) [inline lib/foreman_deployments/dynflow_output_patches.rb:19]: @iNecas could you please check the code from dynflow perspective? I know there was some work planned around active record de-serialization. Has there been any progress in that area? Is this a correct temporary solution or are there better ways? Thanks

**Comment by tstrachota** [comment]
@ShimShtein thanks, it's an important addition!
I briefly read through the code and asked Ivan for checking the dynflow patch part. I haven't tested it yet though.

File: app/lib/foreman_deployments/tasks/search_task_definition.rb (hunk 1 above)
Comment (ShimShtein) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: It's the method that is used by dynflow patch - basically it will be called on deserialization process of dynflow task parameters. First argument (`storage`) is the serialized task output and the second one is a path to the relevant property in the output of the task.
Let's wait for Ivan's comments on this, if he approves the solution I'll add tests.

Comment (iNecas) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: I understand the need for bening able to reference things that are not just hashes and arrays. My concern with the current implementation is however, that it handles deserialization directly in the referencing. I think the `input`/`output` proxy objects might be more smart that they are right now (just plain hashes) and I'm willing to give you a coding hand in making the change in Dynflow, as it would make dynflow more usable. The dereferencing part would that became trivial. What do you think about that approach?

Comment (ShimShtein) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: @iNecas, basically sounds good. Let me tell you my assumptions:
1. Each task knows how to serialize/deserialize its own objects.
2. Dynflow framework doesn't care how the recipient task will handle the output of the sender task.
3. Dynflow needs to supply a stub for the output object in planning phase in order to understand the dependency between the tasks.
4. Dynflow doesn't want to limit the interface between tasks to specific object types (like hash, array and primitives).

Am I missing something?

Comment (iNecas) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: 1. the tasks would specify the implementation to be used for the input/output object, to achieve the extendability.
   I assume one could define a default implementation in configuration. For example, the active record serializer/deserializer should be generic enough for majority of the Rails project.
2. I'm not sure I understand this part. What would be an example of dynflow caring about how the recipient task handles the output
3. yes, this part would stay the same as it's right now
4. no, the only constraint would be, that the `input/output` serializers know how to serialized/deserialize the objects to/from database.

Do this answers fit into you assumptions?

Comment (ShimShtein) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: 1. We agree on this one, which is great. I am talking about the most pessimistic case for what you have mentioned - each task supplies its own unique serializer. Of course it doesn't have to do it - it can reuse a generic serializer supplied by the framework that will know how to serialize common objects like AR, hash, array and primitives. We can probably build the serializer extensible enough so it would be possible to add more objects later on.
2. Let's say sender outputs a single instance of SenderOutputProprietaryClass. It's receiver's responsibility to know that it is possible to call `:foo` on it, and it will return a hash with specific keys. In current implementation, Dynflow assumes that the output is a hash, so it tries to "record" all the keys chain that are used by the recipient.
3. I am not sure in which phase dynflow extracts the dependency. If the output is an arbitrary object, it will be very difficult to mimic its behavior. Besides that I am not sure we actually need to... We need only the fact that the recipient is using the sender.
4. Great, as long as task developers can add/modify the serializers.

Comment (iNecas) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: 1. ack
2. ack - it's the app developer's responsibility to use sensible keys
3. dynflow extracts the dependencies in after planning: we are using the fact, that the output is not computed in planning, so the output object there is just a stub that is used for recording the dependencies: if you tried to use `action.output[:one][:two][three]` in planning, it would give you just an output reference, and this one is replaces by specific data when the run method of the receiver starts (after the sender's run is evaluated and real output is filled in)
4. ack

It seems we are in good agreement on the expectations, I hope to get soon to a draft implementation of this in Dynflow, doesn't look that hard right now (the devil is in details)

Comment (ShimShtein) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: 1. That means the user can't use actual output in the planning phase. It means dynflow has the liberty to forbid actual calls to other tasks. All it has to supply is an interface to say that current task depends on other one. Example for such approach:

``` Ruby
#current state
def plan()
  x = other_task.output[:foo][:get_number] #x will be set to an OutputReference.
  # the fact that this task has a dependency on other_task has been recorded.
  y = x + 1 #will fail, since x is not a number at this point.
end

# suggested state
def plan()
  other_task_state = other_task.result #other_task_result will be set to an OutputReference
  #or maybe even
  other_task_state = add_dependency(:other_task)

  # the fact that this task has a dependency is recorded
  other_task_state.foo # is forbidden - it's a useless call here, since the result will be available only in run phase
end
```

This way we eliminate the need to record call chains for future tasks.

Comment (iNecas) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: We still need to be able to record the chains, because the input of the dependent task doesn't have to match exactly the output of it's dependency. For examples, let's have an action, that produces a host, and other action, that has just an ip on it's input:

```
  host_creation = plan_action(CreateHost, host_attrs)
  plan_action(DhcpReserve, host_creation.output[:mac])
```

I do think the chains make the concept more flexible and allows to combine the actions easier. i don't see a big difference in `output` vs. `result` in planning, we could introduce the `result` but I'm not sure it would help anything.

Comment (ShimShtein) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: Let's begin with the fact that I think that we need to get rid of chain recording, if we want to allow custom output from a task. The reason for this is that if we record the chain, we are responsible to apply it, and this is not always trivial/possible.
The example you are presenting here is very interesting. First, I assume that the snippet belongs to an enveloping task (something like CreateClusterTask), to be more specific - to the planning phase of it.
Let's analyze this example to more details:
1. There is data transformation here: `host_creation.output[:mac]`
2. I see an unusual task ordering here - plan envelope task -> plan CreateHostTask -> **execute CreateHostTask**-> plan DhcpReserve task -> execute DhcpReserveTask

Let's start with [1]: here we can see a data transformation from the output of host creation to input of dhcp reservation. You have described a very simple transform but it doesn't have to be this way. What if the transformation needed here is something like `host_creation.output[:mac].to_binary_array`, or even in a more problematic case, the transformation is not defined on the object itself, something like this: `convert_to_msb_format(host_creation.output[:mac])`. I think the transformation is responsibility of the enveloping task in this example. We sure need to provide a helper that will make the code easier for the user, but recording doesn't solve this. I would suggest something like this:

``` Ruby
  plan_action(DhcpReserve, transform(host_creation) { |output| output.mac.to_binary_array } )
```

under the hood it will create a reference class that will save the delegate and apply it when the output will be ready.
This leads me to issue [2]: Since DhcpCreation task does not have assume some kind of dependency between tasks, it expects plain mac address. This data will be available only after CreateHost task will finish its execution phase. Basically that means that the task scheduling will be similar to [thread pool](https://msdn.microsoft.com/en-us/library/3dasc8as.aspx). The same goes for the enveloping task itself - let's say you want to plan different tasks based on some information of any other task: 

``` Ruby
  host_creation = plan_action(CreateHost, host_attrs)
  if host_creation.output[:status] == :success
    plan_action(DhcpReserve, host_creation.output[:mac])
  else
    plan_action(DeleteHost, host_attrs)
  end
```

You can continue the plan phase only after you have the result of the host creation.

Comment (iNecas) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: Inetersting thoughs, However keep in mind, that the plan needs to happen _before_ the execution starts (that's the whole point of the planning mechanism). If you want to do this kind of decisions, we would need a new concept and I would strictly not recommend trying to add that in to the planning phase. If there is some need for advanced processing of the output, it would need to happen in the `run` phase and I can imaging helpers/concepts to make it easier. But by definition, it can't be done i plan phase.

Comment (ShimShtein) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: OK, the use case with a condition that I have described is indeed very advanced. We can disallow it (at least for now), but what about the original problem that I describe in (2)?

> I see an unusual task ordering here - plan envelope task -> plan CreateHostTask -> execute CreateHostTask-> plan DhcpReserve task -> execute DhcpReserveTask

Comment (tstrachota) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: I like the idea of having an object that takes care of (de)serialization of actions' inputs/outputs. There's an issue open for that in deployments http://projects.theforeman.org/issues/12172 but it would certainly be cleaner to have it supported directly in dynflow.

Regarding @ShimShtein's thoughts in the latest comment it might be interesting to introduce the concept of transformations. Since `OutputReference` takes care of dereferencing itself (correct me if I'm wrong) I think we could add support for using transform functions and have the both concepts could live side by side. Action writers can then choose what's more appropriate for them.

The latest "thread-pool-like" idea seems quite against the dynflow's design to me. As @iNecas mentions, planning heeds to happen before the execution.
BTW specifically in the example with the host I think the `DeleteHost` action should rather be part of the planned rollback mechanism (https://github.com/Dynflow/dynflow/issues/87).

Comment (ShimShtein) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: @tstrachota, Transformations are already a necessity. Right now it's pretty limited and implemented implicitly by square brackets (the transform function is a filter transform - you can only choose a property). After introducing custom output objects, it will be impossible to maintain same transformation code in framework, since the input will become unknown.
I am not a big fan of thread pool either and I don't think it's the ultimate solution. Still we have to address at least the problem with passing (transformed) output as a parameter to plan phase of another task. It is possible to finish the plan phase of the envelope task before the plan phase of dependent task starts. The difficulty here is that planning of tasks that rely on other task's output can be done only after the run phase of the dependency task has  returned with a result.

The example was a stupid one, I just couldn't think of a good one with a condition. Will this example be better:

``` Ruby
  host_creation = plan_action(CreateHost, host_attrs)
  if host_creation.output[:environment] == 'production'
    plan_action(AddServerToSplunkMonitoring, host_creation.output[:ip])
  end
```

Comment (iNecas) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: ----- Original Message -----

> > @@ -7,6 +7,26 @@ def run
> >            results = SerializableArray.new(results)
> >            SearchTaskDefinition.create_output(results, output)
> >          end
> > +
> > -        def self.dereference_output(storage, path)
> 
> OK, the use case with a condition that I have described is indeed very
> advanced. We can disallow it (at least for now), but what about the original
> problem that I describe in (2)?
> 
> > I see an unusual task ordering here - plan envelope task -> plan
> > CreateHostTask -> execute CreateHostTask-> plan DhcpReserve task ->
> > execute DhcpReserveTask

No, it's still `plan CreateHost -> plan DhcpReserve -> execute CreateHost -> execute DhcpReserve`
It's not intended to do much with the inputs in the plan phase: the plan phase
is really mainly for planning and validation if possible.

Comment (iNecas) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: ----- Original Message -----

> In app/lib/foreman_deployments/tasks/search_task_definition.rb :
> 
> > @@ -7,6 +7,26 @@ def run
> >            results = SerializableArray.new(results)
> >            SearchTaskDefinition.create_output(results, output)
> >          end
> > +
> > -        def self.dereference_output(storage, path)
> 
> @tstrachota , Transformations are already a necessity. Right now it's pretty
> limited and implemented implicitly by square brackets (the transform
> function is a filter transform - you can only choose a property). After
> introducing custom output objects, it will be impossible to maintain same
> transformation code in framework, since the input will become unknown.
> I'm don't follow. Can't be just used the approach you just did, with calling a method on the object, if it responds to it 
> and it's not hash? I'm sure it will not work for everything, but the 80-20 rule might apply here quite nicely. 

I think the transformations are nice idea but not really a blocker here. Let me make it straight: getting 
rid of the output sub-chains in planning is no-go here, as it's already being used by Dynflow users. 
We can talk about implementing it better way once the transformations are there, but not sooner. 

> I am not a big fan of thread pool either and I don't think it's the ultimate
> solution. Still we have to address at least the problem with passing
> (transformed) output as a parameter to plan phase of another task. It is
> possible to finish the plan phase of the envelope task before the plan phase
> of dependent task starts. The difficulty here is that planning of tasks that
> rely on other task's output can be done only after the run phase of the
> dependency task has returned with a result.
> 
> The example was a stupid one, I just couldn't think of a good one with a
> condition. Will this example be better:
> host_creation = plan_action( CreateHost , host_attrs) if
> host_creation.output[ :environment ] == ' production ' plan_action(
> AddServerToSplunkMonitoring , host_creation.output[ :ip ]) end

Again: the contract in planning is, that the output is meant just for the planning purposes (passing as input to others 
task). So we are not aiming to support this example. If there is a better name for the output that would 
better describe this contract, I'm ok with introducing alias (but keeping the sub-key chaining).

Comment (ShimShtein) [inline app/lib/foreman_deployments/tasks/search_task_definition.rb:11]: > No, it's still `plan CreateHost -> plan DhcpReserve -> execute CreateHost -> execute DhcpReserve`
> It's not intended to do much with the inputs in the plan phase: the plan phase
> is really mainly for planning and validation if possible.

In this case, DhcpReserve's `#plan` method will get a TaskReference object as a parameter. It's a bit non-trivial for the implementers, but as long as it's well documented, I'm OK with it.

> getting rid of the output sub-chains in planning is no-go here, as it's already being used by Dynflow users.

IMHO we can use them both - we will just use different access points to the output. For example we can leave the `#output` method as is (with chain recording) and create another method `#original_output` that will return the deserialized version of the object (without the chaining).

> I'm ok with introducing alias (but keeping the sub-key chaining)

That's basically what my patch does. If we move the implementation of `self.dereference_output` to `Actions::Base` class, it will become the default behavior, but still be overridable, if the task wishes to. Probably I will break it to two methods - one that deals with root deserializing and one for dealing with call chains. Then we would be able to get rid of the ugly `if` that I add to `ExecutionPlan::OutputReference#dereference`.
Basically we will just move the deserialization responsibility to the task itself - which anyway sounds reasonable to me.
What do you say @iNecas, does the last paragraph sound like a good starting point?
